shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -399,136 +433,447 @@ class Context {
+          this.setToTrueStack.push({ grainId: token.grainId,
+                                     vertexId: vertexIdOfTokenOwner(token),
+                                     permissionId: current.permissionId,
+                                     responsibleTokenId: tokenId, });
+        } else {
+          this.unmetRequirements.addRequirements(token.requirements);
+        }
+      });
+
+      variable.requirementsTailList.forEach((tokenId) => {
+        const token = this.tokensById[tokenId];
+        const activeToken = this.activeTokens[tokenId];
+        activeToken.decrementRequirements(current.permissionId);
+        if (activeToken.requirementsAreMet()) {
+          activeToken.forEachPermission((permissionId, status) => {
+            if ("met" in status) {
+              // We've triggered a new edge! Push it onto the queue.
+              this.setToTrueStack.push({ grainId: token.grainId,
+                                        vertexId: vertexIdOfTokenOwner(token),
+                                        permissionId: permissionId,
+                                        responsibleTokenId: tokenId, });
+            }
+          });
+        }
+      });
+    }
+
+    return this.getPermissions(grainId, vertexId);
+  }
+
+  activateRelevantTokens(grainId, vertexId) {
+    // Returns true if more computation might yield more progress.
+
+    check(grainId, String);
+    check(vertexId, String);
+
+    let result = false;
+    const relevant = computeRelevantTokens(this, grainId, vertexId);
+    this.activateOwnerEdges(grainId, relevant.ownerEdges);
+    relevant.tokenIds.forEach((tokenId) => {
+      if (this.activateToken(tokenId)) {
+        result = true;
+      }
+    });
+
+    return result;
+  }
+
+  processUnmetRequirements(db) {
+    // Returns true if more computation might yield more progress.
+
+    const grainIds = this.unmetRequirements.getGrainIds()
+          .filter((grainId) => !(grainId in this.grains));
+
+    if (db) {
+      this.addGrains(db, grainIds);
+    }
+
+    let result = false;
+
+    while (!this.unmetRequirements.isEmpty()) {
+      const next = this.unmetRequirements.pop();
+      if (this.activateRelevantTokens(next.grainId, "i:" + next.identityId)) {
+        result = true;
+      }
+    }
+
+    return result;
+  }
+
+  tryToProve(grainId, vertexId, permissionSet, db) {
+    // Tries to prove that `vertexId` has the given permissions on the given grain. Returns a
+    // `PermissionSet` representing the permissions proven, or null if it has not been proved
+    // yet that the vertex even has access to the grain.
+
+    check(grainId, String);
+    check(vertexId, String);
+    check(permissionSet, PermissionSet);
+    check(db, Match.OneOf(undefined, SandstormDb));
+    // If `db` is not provided, then this function will make no database queries.
+
+    if (db) {
+      this.addGrains(db, [grainId]);
+    }
+
+    this.activateRelevantTokens(grainId, vertexId, permissionSet);
```

Comment:
Incidentally it seems useful for it to take permissionSet as an argument, too, especially when checking requirements. E.g. if we have a requirement that Bob has write access then we might waste time following tokens that only provide read access. But it also seems like this would be more complicated to implement since now you need partially-activated tokens, or something.